File: day01.js
const fs = require('fs');

let input = fs.readFileSync('inputs/day01.txt', {encoding: 'utf-8'})
  .split('\n')
  .map(line => line.split(/\s+/))
  .map(line => line.map(num => Number(num)));
let left = input
  .map(line => line[0])
  .toSorted((a, b) => a - b);
let right = input
  .map(line => line[1])
  .toSorted((a, b) => a - b);
let diff = 0;
for (let i = 0; i < left.length; i++) {
  diff += Math.abs(left[i] - right[i]);
}
console.log(`Part 1: ${diff}`);

let simScore = 0;
for (const lNum of left) {
  let count = 0;
  for (const rNum of right) {
    if (lNum === rNum) count++;
  }
  simScore += lNum * count;
}
console.log(`Part 2: ${simScore}`);

File: day01.lua
require("util")

local lines = lines_from("inputs/day01.txt")
local left = {}
local right = {}
for _, line in ipairs(lines) do
    local numbers = {}
    for num in string.gmatch(line, "%S+") do
        numbers[#numbers + 1] = tonumber(num)
    end
    table.insert(left, numbers[1])
    table.insert(right, numbers[2])
end
table.sort(left)
table.sort(right)

local diff = 0
for i=1,#left do
    diff = diff + math.abs(left[i] - right[i])
end
print("Part 1: "..diff)

local sim_score = 0
for _, num in ipairs(left) do
    local count = 0
    for _, r in ipairs(right) do
        if r == num then
            count = count + 1
        end
    end
    sim_score = sim_score + num * count
end
print("Part 2: "..sim_score)
